### ISSUE TITLE
Using `failure` with `#![termination_trait]` (`fn main() -> Result<T>`)

### ISSUE THREAD

**Issue by U007D** [withoutboats/failure#130]
I'm running into a couple of issues trying to use `failure` with `#![termination_trait]`.

```
#![feature(termination_trait)]

extern crate failure;

use std::io::{Error as IoError, ErrorKind as IoErrorKind};
use failure::{Fail, Error as FailureError};

type Result<T> = std::result::Result<T, Fail>;

fn main() -> Result<()> {
    Err(IoError::from(IoErrorKind::InvalidData))?
}
```
results in:
```error[E0277]: the trait bound `failure::Fail + 'static: std::marker::Sized` is not satisfied
  --> src/main.rs:10:1
   |
10 | / fn main() -> Result<()> {
11 | |     Err(IoError::from(IoErrorKind::InvalidData))?
12 | | }
   | |_^ `failure::Fail + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `failure::Fail + 'static`
   = note: required by `std::result::Result`

error: aborting due to previous error
```

Switching the `main()` declaration to `type Result<T> = std::result::Result<T, FailureError>` yields:
```
error[E0277]: the trait bound `failure::Fail + 'static: std::marker::Sized` is not satisfied
  --> src/main.rs:14:1
   |
14 | / fn main() -> Result<()> {
15 | |     Err(IoError::from(IoErrorKind::InvalidData))?
16 | | }
   | |_^ `failure::Fail + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `failure::Fail + 'static`
   = note: required by `std::result::Result`

error: aborting due to previous error
```
Is there a solution for this?  If not, should this be a `failure` or a `termination_trait` issue?

Many thanks,
U007D

**Comment by withoutboats** [comment]
The issue you're getting is that you've used `Fail` as the error type, which is just a trait. You should use the `Error` type from failure instead, `type Result<T> = Result<T, failure::Error>;`.

I'll leave this open, please report back whether or not you run into more issues, I want to make sure failure works with this feature.

**Comment by U007D** [comment]
Yes, I tried it using failure::Error too.  It gave:
```rust
error[E0277]: the trait bound `failure::Fail + 'static: std::marker::Sized` is not satisfied
  --> src/main.rs:14:1
   |
14 | / fn main() -> Result<()> {
15 | |     Err(IoError::from(IoErrorKind::InvalidData))?
16 | | }
   | |_^ `failure::Fail + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `failure::Fail + 'static`
   = note: required by `std::result::Result`

error: aborting due to previous error
```

**Comment by withoutboats** [comment]
It looks like the termination trait requires that we implement the old error type. This is unfortunate. :-(

**Comment by U007D** [comment]
Indeed.  Although it is still feature-gated.  I'm never sure where is the right place to leave comments, but the termination_trait PR is [here](https://github.com/rust-lang/rust/pull/46479).  Is this issue something that you could weigh in on to help it be compatible with `failure` before stabilizing?

**Comment by bkchr** [comment]
When I finally have some time, I will improve the implementation to support `failure` :)

**Comment by U007D** [comment]
Hi, @bkchr,

Awesome, thank you.  Is this something I could help with?

**Comment by bkchr** [comment]
It's just changing the implementation, I think we should go the way @withoutboats proposed by restricting the type to `Debug`.

**Comment by U007D** [comment]
Ok, I've forked rust and built it for the first time :tada:.

I also ran `./x.py test` before and after the change and got the same results both times (not all tests run because of a linker error; but of the ones that do run (hundreds), none fail).

https://github.com/U007D/rust/commit/b9b2ca8bd04e932a361adff2af813bd0f9f07e54

Assuming this diff looks right to you two, where should I target the PR?  Back to rust-lang/rust/master?  Or are you working from a branch, @bkchr?

**Comment by bkchr** [comment]
@U007D Back to `rust/master`. My recent changes does not include anything you changed :)

**Comment by U007D** [comment]
Done.
